### PR TITLE
fix: add context timeouts to actions

### PR DESCRIPTION
**Pull request by obs-gh-irismcleary**

### CODE REVIEW

File: cmd/aws-snapshot/main.go [hunk 1]
```diff
@@ -34,6 +35,7 @@ func realMain() error {
 		maxConcurrentRequests = flag.Int("max-concurrent-requests", 10, "Maximum concurrent requests")
 		bufferSize            = flag.Int("buffer-size", 100, "Length of buffer for records")
 		manifestFile          = flag.String("manifest-file", "", "Manifest filename")
+		requestTimeoutSecs    = flag.Int("request-timeout", 300, "Timeout per request (in seconds)")
```
Comment (obs-gh-irismcleary) [inline cmd/aws-snapshot/main.go]: cleanup: make this a string, parse it to a duration

File: cmd/aws-snapshot/main.go [hunk 2]
```diff
@@ -72,6 +74,7 @@ func realMain() error {
 				fmt.Println(string(data))
 			}
 		}),
+		RequestTimeout: time.Duration(*requestTimeoutSecs) * time.Millisecond,
```
Comment (obs-gh-irismcleary) [inline cmd/aws-snapshot/main.go]: no hardcode to ms (this was for testing)

File: cmd/aws-snapshot/main.go (hunk 1 above)
Comment (obs-gh-irismcleary) [reply cmd/aws-snapshot/main.go]: default to 0 (no timeout)?

File: pkg/api/runner.go [hunk 3]
```diff
@@ -41,12 +43,26 @@ func pool(fn Recorder, num int) Recorder {
...
+}
+
 // withSemaphore runs up to maxConcurrency Requests at a time.
-func withSemaphore(fns []Request, maxConcurrency int) Request {
+func withSemaphore(fns []Request, maxConcurrency int, requestTimeout time.Duration) Request {
 	if maxConcurrency == 0 {
 		maxConcurrency = len(fns)
 	}
 
+	if requestTimeout == 0 {
```
Comment (obs-gh-irismcleary) [inline pkg/api/runner.go]: if not set, don't impose a timeout?

File: pkg/api/runner.go [hunk 4]
```diff
@@ -63,7 +65,15 @@ func withSemaphore(fns []Request, maxConcurrency int) Request {
 				defer func() {
 					<-sem
 				}()
-				errCh <- fn(ctx, ch)
+
+				if requestTimeout == 0 {
+					errCh <- fn(ctx, ch)
```
Comment (jta) [inline pkg/api/runner.go]: Is there any way of making it so we only have one place where we write to the channel? Having multiple branches is harder to debug / maintain over time.

File: pkg/api/runner.go [hunk 5]
```diff
@@ -12,6 +13,7 @@ type Runner struct {
 	BufferSize            int
 	MaxConcurrentRequests int
 	ConcurrentRecorders   int
+	RequestTimeout        time.Duration
```
Comment (jta) [inline pkg/api/runner.go]: Alternatively you could define as a pointer `*time.Duration`. This has the advantage that it is less ambiguous around whether a value is needed or not. Your current approach works fine, it's just odd to see that the default in command line is `0s` (users may get confused as to what that even means).

File: pkg/api/runner.go (hunk 4 above)
Comment (obs-gh-irismcleary) [reply pkg/api/runner.go]: we have to pass different contexts for the timeout vs no-timeout case, assuming preserving no-timeout is desired. If this were Python I could use a ternary statement to pass the context, but it's not 😛. Would it be better or worse to declare var reqErr above, set reqErr in each branch, and then write that to the channel? I suppose another option would be to set cctx to either a new context.WithTimeout or the original parent context, do the branching there, and only have one execution of fn

File: pkg/api/runner.go (hunk 5 above)
Comment (obs-gh-irismcleary) [reply pkg/api/runner.go]: yeah that would make for a clearer cmd line. Will do

File: pkg/api/runner.go (hunk 4 above)
Comment (jta) [reply pkg/api/runner.go]: Something like the following would probably work (assuming you keep `time.Duration` vs `*time.Duration`:
```
if requestTimeout > 0 {
    var cancel context.CancelFunc
    ctx, cancel = context.WithTimeout(ctx, requestTimeout)
    defer cancel()
}
errCh <- fn(ctx, ch)
```

Comment (obs-gh-irismcleary) [reply pkg/api/runner.go]: I'm concerned about that changing what ctx is in the outer withSemaphore fn. It's not passed into the per-req go func, we were implicitly capturing and using the one from outside.

I suppose I could change that, make the inner go func take a request and a context, and then modify the one passed in before actually supplying it to the fn

File: pkg/api/runner_test.go [hunk 6]
```diff
@@ -42,3 +54,27 @@ func TestRunner(t *testing.T) {
...
+		Requests: []api.Request{
+			testLongRequest,
+		},
+		MaxConcurrentRequests: 1,
+		ConcurrentRecorders:   1,
+		Recorder:              &recorder,
+		RequestTimeout:        &timeout,
+	}
+
+	if err := r.Run(context.Background()); err == nil {
```
Comment (jta) [inline pkg/api/runner_test.go]: You can do slightly better here, and verify that the error matches `context.DeadlineExceeded`

File: pkg/api/runner.go [hunk 7]
```diff
@@ -59,12 +61,20 @@ func withSemaphore(fns []Request, maxConcurrency int) Request {
 
 		for _, fn := range fns {
 			sem <- struct{}{}
-			go func(fn Request) {
+			go func(fn Request, ctx context.Context) {
```
Comment (jta) [inline pkg/api/runner.go]: Typically context comes first in list of arguments.

Comment (obs-gh-irismcleary) [reply pkg/api/runner.go]: heh, facepalm

File: pkg/api/runner_test.go [hunk 8]
```diff
@@ -42,3 +44,38 @@ func TestRunner(t *testing.T) {
...
+				}
+			},
+		},
+		Recorder:       &recorder,
+		RequestTimeout: &timeout,
+	}
+
+	err := r.Run(context.Background())
+
+	if err == nil {
```
Comment (jta) [inline pkg/api/runner_test.go]: This is not needed, it'd get caught by your next check

**Review by jta — APPROVED** [review]

**Pull request merged**